docs/guides/versioning.md

```diff
@@ -0,0 +1,312 @@
+
+We realize this can be painful for consuming projects. In attempt to ease this pain, here are a few supporting thoughts you may find helpful:
+
+1. Overriding styles from this library in your application is a valid use case that we would like to assist in supporting. While there is no way for this library to test your codebase for overriden styles, [version release changelogs](https://github.com/IBM/carbon-addons-iot-react/releases) are provided for each release, which contain a list of included commits and recently closed issues. This can be used to identify high level potential changes when upgrading versions of this library in your application. Additionally you may consider writing automated tests against styles you override to ensure their validity when updating.
+1. Changes to the public component API is tracked via [a jest snapshot file](https://github.com/IBM/carbon-addons-iot-react/blob/master/src/utils/__tests__/__snapshots__/publicAPI.test.js.snap). This includes `className` and other prop changes. The file can be used during your upgrade process to compare changes between different versions of the library. [Here is an example comparing v2.55.0 -> v2.56.0](https://github.com/IBM/carbon-addons-iot-react/compare/v2.55.0...v2.56.0#diff-789e184fd0335570057a913379ee5f43), tags can be changed and selected via the dropdowns at the top of that page.
+1. For testing automation, we are working to provide a dedicated `data-testid` data attribute placed on the containing element of each component. This should be used in place of class or id selectors in your testing automation. More info is available in the [testing guide documentation](/testing).
+
+#### A `data-testid` attribute is added
+
+semver bump: **minor**
+
+```diff
+- function ExportedComponent({ message }) {
++ function ExportedComponent({ testID, message }) {
+  return (
+    <>
+-     <span>{message}</span>
++     <span data-testid={testID}>{message}</span>
+    </>
+  );
+}
+
+ExportedComponent.propTypes = {
+  message: PropTypes.string,
++ testID: PropTypes.string,
+};
+ExportedComponent.defaultProps = {
+  message: '',
++ testID: 'ExportedComponent',
+};
+```
+
+#### A `data-testid` attribute default prop value is changed or removed
+
+semver bump: **major**
+
+```diff
+function ExportedComponent({ testID, message }) {
+  return (
+    <>
+      <span>{message}</span>
+      <span data-testid={testID}>{message}</span>
+    </>
+  );
+}
+
+ExportedComponent.propTypes = {
+  message: PropTypes.string,
+  testID: PropTypes.string,
+};
+ExportedComponent.defaultProps = {
+  message: '',
+- testID: 'ExportedComponent',
++ testID: 'ExportedComponentNewName',
+};
+```
+
+#### The DOM node that an `id` corresponds to is changed
+
+semver bump: **minor**
```

Comment:
I argue that in the event that is no existing data-test-id then this is a major change and would be like removing a data-test-id

Reply:
@stuckless I'm not opposed, but Carbon doesn't consider this a breaking change. So if we pull in a new minor version of Carbon and the DOM node for an ID has changed, we would be forced to issue a new major/breaking version.

If we kept this as a minor change, we could mitigate the problem by working to ensure the DOM nodes for IDs are included in a snapshot, so we can provide accurate release/migration notes for when these things change. Thoughts?

Reply:
Yeah, I'm just trying to show the difference in the meaning of "breaking" when it comes to APIs vs UI LIbraries.   Carbon (and yourself) really are only looking at the API side of things, but there are lots of things that break an application from the UI side.  And moving the ID to a new element would be one of them. (ie, the app side has a test selector based off the id of the component and now all the tests fail because the id moved to a different location)   Now, if this is a case where there is an existing data-test-id then I agree, we should be using the data-test-id and moving the id should be minor.   But the same logic would apply to moving a data-test-id to a new element, and that would be a breaking change.

Reply:
👍 I've made a note of this in #1024